File: Python/programmers/etc/level2/우박수열_정적분.py
def solution(k, ranges):
    answer = []
    
    hail = []
    while k > 1:
        tmp = k
        if k&1:
            k = k*3+1
        else:
            k = k//2
        hail.append((tmp+k)/2)
    
    dp = [0]*(len(hail)+1)
    for i in range(1, len(hail)+1):
        dp[i] = dp[i-1] + hail[i-1]
    
    for a, b in ranges:
        b = len(dp)+b-1
        if a > b:
            answer.append(-1.0)
            continue
        answer.append(dp[b]-dp[a])
        
    return answer


k = 5
ranges = [[0,0],[0,-1],[2,-3],[3,-3]]
print(solution(k, ranges)) # [33.0,31.5,0.0,-1.0]

File: Java/programmers/etc/level2/우박수열정적분.java
package CodeTest.Java.programmers.etc.level2;

import java.util.*;

public class 우박수열정적분 {
    public double[] solution(int k, int[][] ranges) {
        List<Integer> hail = new ArrayList<>();
        
        while (k>1) {
            hail.add(k);
            if (k%2 == 1) {
                k = k*3+1;
            } else {
                k /= 2;
            }
        }
        hail.add(k);
        
        double[] answer = new double[ranges.length];
        
        for (int i=0; i<ranges.length; i++) {
            int a = ranges[i][0], b = ranges[i][1];
            if (a > b+hail.size()-1) {
                answer[i] = -1;
                continue;
            } else if (a == b+hail.size()-1) {
                answer[i] = 0;
                continue;
            }
            
            double dimension = 0;
            for (int j=a; j<b+hail.size()-1; j++) {
                dimension += (hail.get(j)+hail.get(j+1))/2.0;
            }
            
            answer[i] = dimension;
        }
        
        return answer;
    }
}
